src/storage.rs

```diff
@@ -1,296 +1,494 @@
-        let cache_valid = if self.reader_internal.len() > self.current_longest_reader {
-            let cached_reader =
-                unsafe { &*self.reader_internal[self.current_longest_reader].get() };
-            cached_num_written = if self.written < cached_reader.written {
-                self.written + (self.reset_written - cached_reader.written)
-            } else {
-                self.written - cached_reader.written
-            };
-            cached_reader.alive.load(Ordering::Relaxed) && cached_num_written > 1
-        } else {
-            false
-        };
-        if cache_valid {
-            cached_num_written > self.data.len()
-        } else {
-            let (longest_reader, num_written) = self.reader_internal
-                .iter()
-                .map(|ref internal| unsafe { &*internal.get() })
-                .enumerate()
-                .filter_map(|(i, internal)| {
-                    if internal.alive.load(Ordering::Relaxed) {
-                        Some((i, &internal.written))
-                    } else {
-                        None
-                    }
-                })
-                .fold((0, 0), |(index, max_written), (i, &written)| {
-                    let num_written = if self.written < written {
-                        self.written + (self.reset_written - written)
-                    } else {
-                        self.written - written
-                    };
-                    if num_written > max_written {
-                        (i, num_written)
-                    } else {
-                        (index, max_written)
-                    }
-                });
-            self.current_longest_reader = longest_reader;
-            num_written > self.data.len()
+    /// Ensures that `num` elements can be inserted.
+    /// Does nothing if there's enough space, grows the buffer otherwise.
+    #[inline(always)]
+    pub fn ensure_additional(&mut self, num: usize) {
+        if self.available >= num {
+            return;
         }
+
+        self.ensure_additional_slow(num);
     }
 
-    /// Write a single data point into the ringbuffer.
-    pub fn single_write(&mut self, data: T) {
-        self.written += 1;
-        if self.written > self.reset_written {
-            self.written = 0;
-        }
-        let need_growth = self.needs_growth();
-        if need_growth {
-            self.data.insert(self.write_index, data);
-            // In order to avoid pushing events that have already been read back into a readable
-            // range we're also going to push any readers that are ahead of us in the buffer
-            // forward one as well.
-            for i in 0..self.reader_internal.len() {
-                unsafe {
-                    if (*self.reader_internal[i].get()).index > self.write_index {
-                        (*self.reader_internal[i].get()).index += 1;
-                    }
-                }
+    #[inline(never)]
```

Comment:
I just read on #inline and I think this is not required and would slow the program? Is it the objective? "ensure_additional_SLOW"

Reply:
Not sure what you mean exactly.